### DISCUSSION TITLE
performance profiling

### DISCUSSION THREAD

**Discussion by hw-ju** [Project-MONAI/tutorials#675] [Q&A] [answered]
Hi! Are `with Range("XXX"):` and `with nvtx.annotate("XXX"):` essentially the same thing? 
Thanks!

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

Could you please help share some comments about this question?

Thanks in advance.

**Comment by bhashemian** [comment, marked as answer]
Hi @hw-ju,

Thanks for your question. They are not exactly the same but very similar. They both define the span of an event but there are two differences:

1. `with Range` uses `torch.cuda.nvtx` and it only depends on the cuda version of PyTorch. However, `with nvtx.annotate` uses `nvtx` library directly, so it depends on `nvtx` being installed.
2. `with Range` uses [`range_push`](https://pytorch.org/docs/stable/generated/torch.cuda.nvtx.range_push.html#torch.cuda.nvtx.range_push) and [`range_pop`](https://pytorch.org/docs/stable/generated/torch.cuda.nvtx.range_pop.html#torch.cuda.nvtx.range_pop) on the backend to define the context, so it inherit their limitations like they currently does not support to change the range color, while in [`with nvtx.annotate`](https://nvtx.readthedocs.io/en/latest/basic.html#annotate) you can set the color in the input argument.

**Reply by hw-ju** [reply]
Very helpful! Thanks @drbeh! :D
